File: submissions/0323-number-of-connected-components-in-an-undirected-graph/solution.ts
function countComponents(n: number, edges: number[][]): number {
   
  let graph = new Map();  
  let seen = new Array(n).fill(false);
    
  for(let i = 0; i < n; i++){
      graph.set(i, []);
  }
  
  for(const [x, y] of edges){
      graph.get(x).push(y);
      graph.get(y).push(x);
  }  
    
  let dfs = node => {
      for(const neighbour of graph.get(node)){
          if(!seen[neighbour]){
              seen[neighbour] = true;
              dfs(neighbour);
          }
      }
  }
  
  let ans = 0;
    
  for(let i = 0; i < n; i++){
      if(!seen[i]){
         ans+=1
         seen[i]=true
         dfs(i); 
      }
  }  
  
  return ans;  
};

File: submissions/0323-number-of-connected-components-in-an-undirected-graph/solution.py
from collections import defaultdict
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        seen = set()
        graph = defaultdict(list)
        
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
        
        def dfs(node):
            for neighbour in graph[node]:
                if neighbour not in seen:
                    seen.add(neighbour)
                    dfs(neighbour)
        
        ans = 0
        for node in range(n):
            if node not in seen:
                seen.add(node)
                dfs(node)
                ans += 1
                
        return ans
